include/dice/template-library/defer.hpp

```diff
@@ -0,0 +1,155 @@
+	 * Inspired by Andrei Alexandrescu's “Declarative Control Flow" presentation
+	 *
+	 * @tparam Pol Policy deciding if the function should be executed
+	 * @tparam F type of function to execute
+	 */
+	template<ScopeExitPolicy Pol, typename F>
+	struct ScopeExitGuard {
+	private:
+		std::optional<F> func_;
+		int const uncaught_exceptions_ = std::uncaught_exceptions();
+
+		[[nodiscard]] bool scope_failed() const noexcept {
+			return uncaught_exceptions_ < std::uncaught_exceptions();
+		}
+
+	public:
+		ScopeExitGuard() noexcept = default;
+
+		explicit ScopeExitGuard(F &&func) noexcept(std::is_nothrow_move_constructible_v<F>) : func_{std::move(func)} {
+		}
+
+		explicit ScopeExitGuard(F const &func) noexcept(std::is_nothrow_copy_constructible_v<F>) : func_{std::move(func)} {
+		}
+
+		ScopeExitGuard(ScopeExitGuard const &) = delete;
+		ScopeExitGuard &operator=(ScopeExitGuard const &) = delete;
+
+		~ScopeExitGuard() noexcept(Pol != ScopeExitPolicy::OnSuccess) {
+			if (!func_.has_value()) {
+				return;
+			}
+
+			if constexpr (Pol == ScopeExitPolicy::OnSuccess) {
+				if (!scope_failed()) {
+					std::invoke(*func_);
+				}
+			} else if constexpr (Pol == ScopeExitPolicy::OnFail) {
+				if (scope_failed()) {
+					std::invoke(*func_);
+				}
+			} else /* Pol == ScopeExitPolicy::Always */ {
+				std::invoke(*func_);
+			}
+		}
+	};
+
+	/**
+	 * Constructs a ScopeExitGuard with ScopeExitPolicy::Always and the given function
+	 * @param func function to execute on scope exit
+	 * @return constructed ScopeExitGuard
+	 */
+	template<typename F>
+	auto make_scope_exit_guard(F &&func) {
+		return ScopeExitGuard<ScopeExitPolicy::Always, std::remove_cvref_t<F>>{std::forward<F>(func)};
+	}
+
+	/**
+	 * Constructs a ScopeExitGuard with ScopeExitPolicy::OnFail and the given function
+	 * @param func function to execute on scope exit
+	 * @return constructed ScopeExitGuard
+	 */
+	template<typename F>
+	auto make_scope_fail_guard(F &&func) {
+		return ScopeExitGuard<ScopeExitPolicy::OnFail, std::remove_cvref_t<F>>{std::forward<F>(func)};
+	}
+
+	/**
+	 * Constructs a ScopeExitGuard with ScopeExitPolicy::OnSuccess and the given function
+	 * @param func function to execute on scope exit
+	 * @return constructed ScopeExitGuard
+	 */
+	template<typename F>
+	auto make_scope_success_guard(F &&func) {
+		return ScopeExitGuard<ScopeExitPolicy::OnSuccess, std::remove_cvref_t<F>>{std::forward<F>(func)};
+	}
+
+	namespace detail {
+		struct ScopeGuardOnExit {};
+		struct ScopeGuardOnFail {};
+		struct ScopeGuardOnSuccess {};
+
+		template<typename F>
+		auto operator+(ScopeGuardOnExit, F &&func) {
+			return make_scope_exit_guard(std::forward<F>(func));
+		}
+
+		template<typename F>
+		auto operator+(ScopeGuardOnFail, F &&func) {
+			return make_scope_fail_guard(std::forward<F>(func));
+		}
+
+		template<typename F>
+		auto operator+(ScopeGuardOnSuccess, F &&func) {
+			return make_scope_success_guard(std::forward<F>(func));
+		}
```

Comment:
Yeah, basically to improve the observable syntax. So that it seems more like part of the language (and this is also probably less error prone).

Error messages are surprisingly really good (example):
```
/home/liss/Development/dice-template-library/tests/tests_defer.cpp:84:11: error: expected body of lambda expression
   84 |                         DEFER();
```

Reply:
added comment to namespace